src/simperium/channel.js

```diff
@@ -643,35 +643,44 @@ LocalQueue.prototype.resendSentChanges = function() {
 }
 
 function collectionRevisions( channel, id, callback ) {
-	var expectedVersions = -1;
-	var onGhostRetrieved = function( ghost ) {
-		var version = Math.min( ghost.version, 30 );
-		var i;
-		expectedVersions = version;
+	let expectedVersions;
+	let timeout;
+	const TIMEOUT = 10000; // arbitrarily chosen delay
 
-		// Loop through requested revision count and request each version
-		for ( i = 0; i < version; i++ ) {
-			channel.send( 'e:' + id + '.' + ( ghost.version - i ) );
-		}
+	const versions = [];
+
+	const finish = listener => {
+		channel.removeListener( 'version.' + id, listener );
+		callback( null, versions.sort( ( a, b ) => a.version - b.version ) );
+		clearTimeout( timeout );
 	};
 
-	var versions = [];
-	var onVersion = function( id, version, data ) {
-		versions.push( {id: id, version: version, data: data} );
+	const onVersion = ( id, version, data ) => {
+		versions.push( { id, version, data } );
 
 		// Check if all versions have been collected
 		if ( expectedVersions === versions.length ) {
-			channel.removeListener( 'version.' + id, onVersion );
-			callback( null, versions.sort( function( a, b ) {
-				return a.version > b.version ? -1 : 1;
-			} ) );
+			finish( onVersion );
 		}
 	};
 
 	channel.on( 'version.' + id, onVersion );
 
-	channel.store.get( id ).then( onGhostRetrieved, function( e ) {
-		callback( e );
-	} );
+	channel.store.get( id ).then( ghost => {
+		// the default bucket options allow for storing
+		// the 60 most-recent revisions of a note plus
+		// 100 archive versions (these store one out of
+		// every ten versions). we'll get up to this many
+		const version = Math.min( ghost.version, 160 );
+		expectedVersions = version;
+
+		// Loop through requested revision count and request each version
+		for ( let i = 0; i < version; i++ ) {
+			channel.send( 'e:' + id + '.' + ( ghost.version - i ) );
```

Comment:
Will this start requesting versions that are no longer stored once `i > 60`?

Reply:
you know what it _will_ do? it will _still_ fail to request all of the revisions. supposing we are on revision 2000, it will only request back to 1840, but in reality, we'll have 1940-2000 in the revision list then 1931, 1921, 1911, …, 931 in the archive list.

I thought about just sending requests back down to revision 1 but then we definitely won't get all of those back

When I was cutting out refactorings to shrink the PR I had eliminated an experiment based on timing. Requests a pool of updates at first then start incrementally requesting the previous batch. After a timeout kill the process.

```js
range( 30 )
	.map( i => ghost.version - i )
	.forEach( v => channel.send( `e:${ id }.${ v }` ) )

onVersion() {
	deferFinish( DELAY );
	versions.push( … );

	if ( versions.length === BATCH_SIZE ) {
		startIterativeFetch();
	}
}

startIterativeFetch( version ) {
	channel.send( `e:${ id }.${ version - 1 }` );
}

finish() {
	callback( null, versions.sort( versionSort ) );
	channel.removeListener( 'version' );
}
```

I could put this together again to try and make it clear, but it essentially wipes out the existing function entirely _and_ moves to a race-condition-based approach. Tragically I couldn't find any way to get the list of available revisions which would dhave completely resolved this issue.

Reply:
> you know what it will do? it will still fail to request all of the revisions. supposing we are on revision 2000, it will only request back to 1840, but in reality, we'll have 1940-2000 in the revision list then 1931, 1921, 1911, …, 931 in the archive list.

My question was somewhat terse, but yes, that's what I was thinking too. Also this might be the answer to our original question - _why was the count set to 30 in the first place_? Maybe they just didn't want to end up handling the skipped ones in archive range. Or maybe at some point in history the default bucket limit for most recent revisions was 30 instead of 60.

I like the batching idea, although I don't think know how the delays in fetching the batches might impact the client.

Side note: all of this would be a lot easier if versions were represented as doubly linked list. :smile:

Reply:
Revisions are available in the database and we can go back until the server sends a `?` - the part I find most difficult is the flow inside of this library. I'll be digging into it.